aequilibrae/paths/cython/route_choice_set.pyx

```diff
@@ -382,6 +383,140 @@ cdef class RouteChoiceSet:
             self.get_sl_link_loading(cores=c_cores)
             self.get_sl_od_matrices()
 
+    def assign_from_df(
+        self,
+        graph: pd.DataFrame,
+        df: pd.DataFrame,
+        demand: GeneralisedCOODemand,
+        select_links: Dict[str, FrozenSet[FrozenSet[int]]] = None,
+        recompute_psl: bool = False,
+        sl_link_loading: bool = True,
+        store_results: bool = True,
+        beta: float = 1.0,
+        cutoff_prob: float = 0.0,
+    ):
+        cdef:
+            long int c_cores = 1  # Single threaded only due to high python interop, this shoud be fast anyway
+            int thread_id = 0
+
+            # Scale cutoff prob from [0, 1] -> [0.5, 1]. Values below 0.5 produce negative inverse binary logit values.
+            double scaled_cutoff_prob = (1.0 - cutoff_prob) * 0.5 + 0.5
+
+        # We need to know the indices of the demand matrix that our OD pairs correspond to
+        try:
+            od_indices = (
+                demand.df.assign(index=np.arange(len(demand.df)))
+            ).loc[
+                df[["origin id", "destination id"]].drop_duplicates().itertuples(name=None, index=False),
+                "index"
+            ].reset_index()
+        except KeyError:
+            raise KeyError("not all origin and destinations IDs from the path files are present within the demand matrix")
```

Comment:
All from the demand matrix MUST exist in the path files, but there is no need to exist demand for all OD pairs in the path file.  Think of the case where people prepare the choice sets for the ENTIRE zone system and then simply apply them to whatever matrix they have